app_legacy/Helpers/database/player-game.php

```diff
@@ -91,60 +91,193 @@ function getGameRankAndScore(int $gameID, string $username): array
+                'ConsoleID' => (int) $gameData['ConsoleID'],
+                'ConsoleName' => $gameData['ConsoleName'],
+                'ForumTopicID' => (int) $gameData['ForumTopicID'],
+                'Flags' => (int) $gameData['Flags'],
+                'ImageIcon' => $gameData['ImageIcon'],
+                'ImageTitle' => $gameData['ImageTitle'],
+                'ImageIngame' => $gameData['ImageIngame'],
+                'ImageBoxArt' => $gameData['ImageBoxArt'],
+                'Publisher' => $gameData['Publisher'],
+                'Developer' => $gameData['Developer'],
+                'Genre' => $gameData['Genre'],
+                'Released' => $gameData['Released'],
+                'IsFinal' => (int) $gameData['IsFinal'],
+            ];
+        }
     }
-    $gameIDs = implode(',', $gameIDs);
-
-    // Count num possible achievements
-    $query = "SELECT GameID, COUNT(*) AS AchCount, SUM(ach.Points) AS PointCount FROM Achievements AS ach
-              WHERE ach.Flags = " . AchievementType::OfficialCore . " AND ach.GameID IN ( $gameIDs )
-              GROUP BY ach.GameID
-              HAVING COUNT(*)>0 ";
-
-    $dbResult = legacyDbFetchAll($query);
-    foreach ($dbResult as $data) {
-        $dataOut[$data['GameID']]['NumPossibleAchievements'] = $data['AchCount'];
-        $dataOut[$data['GameID']]['PossibleScore'] = $data['PointCount'];
+    $libraryOut['Awarded'] = $awardedData;
+
+    if ($withGameInfo) {
+        $libraryOut['GameInfo'] = $gameInfo;
     }
 
-    // Foreach return value from this, cross-reference with 'earned' achievements. If not found, assume 0.
-    // Count earned achievements
-    $query = "SELECT GameID, COUNT(*) AS AchCount, SUM( ach.Points ) AS PointCount, aw.HardcoreMode
-              FROM Awarded AS aw
-              LEFT JOIN Achievements AS ach ON aw.AchievementID = ach.ID
-              WHERE ach.GameID IN ( $gameIDs ) AND ach.Flags = " . AchievementType::OfficialCore . "
-              AND aw.User = :username
-              GROUP BY aw.HardcoreMode, ach.GameID";
-
-    $dbResult = legacyDbFetchAll($query, ['username' => $user]);
-    foreach ($dbResult as $data) {
-        if ($data['HardcoreMode'] == 0) {
-            $dataOut[$data['GameID']]['NumAchieved'] = $data['AchCount'];
-            $dataOut[$data['GameID']]['ScoreAchieved'] = $data['PointCount'];
-        } else {
-            $dataOut[$data['GameID']]['NumAchievedHardcore'] = $data['AchCount'];
-            $dataOut[$data['GameID']]['ScoreAchievedHardcore'] = $data['PointCount'];
+    if ($numRecentAchievements >= 0) {
+        usort($unlockedAchievements, function ($a, $b) {
+            if ($a['When'] == $b['When'])
+                return $a['Achievement']['ID'] <=> $b['Achievement']['ID'];
+
+            return -($a['When'] <=> $b['When']);
+        });
+
+        if ($numRecentAchievements !== 0) {
+            $unlockedAchievements = array_slice($unlockedAchievements, 0, $numRecentAchievements);
+        }
```

Comment:
this represents the old limit (which was 5000 when called from the user's profile, or 10 (by default) when called through API_GetUserSummary). Instead of limiting the query to the last N unlocks, we merge the unlocks from the recent games, then truncate the list at N.